ducktape/tests/test.py

```diff
@@ -32,17 +33,85 @@ def __init__(self, test_context):
+    def tearDown(self):
+        """Default teardown method which stops and cleans services registered in the ServiceRegistry.
+        Note that there is not a default setUp method. This is because the framework makes no assumptions
+        about when and in what methods the various services are started.
+        """
+        if hasattr(self, 'services') and type(self.services) == ServiceRegistry:
+            self.services.stop_all()
+            self.copy_service_logs()
+            self.services.clean_all()
+
+    def free_nodes(self):
+        self.services.free_all()
+
+    def copy_service_logs(self):
+        """Copy logs from service nodes to the results directory."""
+        for service in self.services.values():
+            if not hasattr(service, 'logs'):
+                self.test_context.logger.debug("Won't collect service logs from %s - no 'logs' attribute." %
+                    service.__class__.__name__)
+                return
+
+            log_dirs = service.logs
+            for node in service.nodes:
+                for log_name in log_dirs.keys():
+                    if self.should_collect_log(log_name, service, node):
+                        dest = os.path.join(
+                            self.test_context.results_dir, service.__class__.__name__, node.account.hostname)
+
+                        if not os.path.isdir(dest):
+                            mkdir_p(dest)
+
+                        try:
+                            node.account.scp_from(log_dirs[log_name], dest, recursive=True)
+                        except Exception as e:
+                            self.test_context.logger.warn(
+                                "Error copying log %(log_name)s from %(source)s to %(dest)s. \
+                                service %(service)s: %(message)s" %
+                                {'log_name': log_name,
+                                 'source': log_dirs[log_name],
+                                 'dest': dest,
+                                 'service': service,
+                                 'message': e.message})
+
+    def mark_log_for_collect(self, log_name, service, node=None):
+        if node is None:
+            # By default, mark all nodes for collection
+            for n in service.nodes:
+                self.log_collect[(log_name, service.__class__.__name__, n.account.hostname)] = True
+        else:
+            self.log_collect[(log_name, service.__class__.__name__, node.account.hostname)] = True
+
+    def mark_log_for_no_collect(self, log_name, service, node):
+        if node is None:
+            # By default, mark all nodes for collection
+            for n in service.nodes:
+                self.log_collect[(log_name, service.__class__.__name__, n.account.hostname)] = False
+        else:
+            self.log_collect[(log_name, service.__class__.__name__, node.account.hostname)] = False
+
+    def should_collect_log(self, log_name, service, node):
```

Comment:
So if I'm interpreting this right, the model is that Services specify all logs that a test might be interested in keeping, and tests blacklist anything they don't want? If we're doing this for Kafka, do we always want opt-in? I'm wondering if defaulting to always collecting the data logs (which might be very large) and having to repeat the filtering logic in every test that doesn't need them is a good idea?